Text editor documentation (#9743)
godotengine/godot-docs: tutorials/editor/script_editor.rst

.. |documentation| image:: img/script_editor_icons/Documentation.webp
.. |toolscript| image:: img/script_editor_icons/ToolScript.webp
	
.. image:: img/editor_ui_script_editor_script_panel.webp

Below the menus, on the left panel, you will see a list of opened files and documents
if you have any. Depending on the file type, this list will have an icon next
to the file name. For example, the |script| icon means that it is a GDScript.
the |scriptcsharp| means it is a C# script. The |documentation| means that this is a 
built-in class reference. Finally, the |toolscript| means it is a currently running 
script (See :ref:`tool annotation <doc_running_code_in_the_editor>` for more on this). 
Hovering a file will show a tooltip to its relative location in the project folder.

On the status bar, clicking the left arrow hides the script panel, clicking
on the right arrow shows it.

If you did not change any settings, the file names may also have a different coloring.
This helps you identify the recently edited files by highlighting them. This behavior
can be changed in the `Editor > Editor Settings` by adjusting the "Script Temperature"
properties in the "Text Editor" section.

> The filter bar above the file names introduces a handy case insensitive search to

Reviewer: ```suggestion
The filter bar above the file names introduces a handy case-insensitive search to
```

> find a specific file. Even if you just type the letters in a file name into the

Reviewer: ```suggestion
find a specific file. Even if you just type the letters of a file name into the
```

bar, files containing these letters in order will also appear. Assume that there
is a file named ``button.gd`` in the list. If you type ``btn`` into the filter bar, 
this file will appear in the results.  To reset the filter, clear the filter bar.

An asterisk (*) next to a file name indicates that the file has unsaved changes.

> .. tip:: If you just put "*" in the filter bar, you can display all unsaved files.

Reviewer: ```suggestion
.. tip:: If you just enter "*" in the filter bar, you can display all unsaved files.
```

You can drag a file to change the ordering. Right-clicking on a file provides several

> options to save or close the files, or to copy the relative path of the file:

Reviewer: ```suggestion
options to save or close files, or to copy the relative path of the file:
```

> You can use the "Move Up", "Move Down" to change the order of the file, or use "Sort"

Reviewer: ```suggestion
You can use "Move Up" and "Move Down" to change the order of the file, or use "Sort"
```

to sort all files alphabetically. "Toggle Scripts Panel" hides the panel, which
can be displayed again using the right arrow on the status bar.

> "Close Docs" options closes all opened in-class reference documents leaving only

Reviewer: ```suggestion
The "Close Docs" option closes all opened in-class reference documents leaving only
```
Or 
```suggestion
"Close Docs" closes all opened in-class reference documents leaving only
```

script files open. "Show in FileSystem" finds and highlights the file in the
FileSystem dock.

> Below the file list, you'll see the name of the currently opened file. The button

Reviewer: ```suggestion
Below the file list, you'll see the name of the currently open file. The button
```

next to this switches the ordering of the methods defined in this file between
alphabetical and as they appear. Under this is the outline of the file. If this
is a script file, it will contain the list of defined methods. If, however, a

> class reference is open, this area will list the table of contents of this

Reviewer: ```suggestion
class reference page is open, this area will show the table of contents of this
```

document. Clicking on an item in this list will jump to the respective function
or section in the file. Similarly, the "Filter Methods" bar lets you search
for a specific function or section within the selected document with the same
behavior as filtering scripts.

.. _doc_script_editor_menus:

Menus
-----

The text editor's menus lie below the scene switcher and allow you to access a
variety of tools and options, such as file management, search and replace, debugging
controls, and code formatting features.

.. tip:: An asterisk (*) next to an action means that this operation is also available
  in the context menu, which can be opened by right-clicking in the code editor.

.. image:: img/script_editor_icons/text_editor_menu.webp

**File** menu provides the following options:

.. image:: img/script_editor_icons/text_editor_file_menu.webp

> - **New Script...**: Opens the new script dialog to create and add the script into

Reviewer: ```suggestion
- **New Script...**: Opens the new script dialog to create and add the script to
```

>   the project. If the creation is successful, it will directly open it in the

Reviewer: ```suggestion
  the project. If creation is successful, it will directly open it in the
```

  text editor. Depending on the  version of Godot (with C# support or not), you

>   can choose ``.gd`` or ``.cs`` as extension.

Reviewer: ```suggestion
  can choose ``.gd`` or ``.cs`` as the extension.
```

- **New Text File...**: Opens the file dialog to create a plain text file with
  one of the recognized formats. Godot can also highlight ``json`` files.
- **Open..**: Opens the file dialog to let you browse inside your computer and
  choose any recognized text file to open it.
- **Reopen Closed Script**: Reopens the last closed scripts. You can use this
  option multiple times to reopen other closed scripts if you closed more than one.
- **Open Recent**: Provides a list of last opened scripts. You can also clear the
  list using the provided option at the bottom of the list.
- **Save**: Saves the currently selected script.

> - **Save As...**: Opens the file dialog to save the currently selected script

Reviewer: ```suggestion
- **Save As...**: Opens the file dialog to save the currently open script
```

  with a different name.

> - **Save All**: Saves all unsaved scripts opened in the text editor. Scripts with

Reviewer: ```suggestion
- **Save All**: Saves all unsaved open scripts in the text editor. Scripts with
```

  unsaved changes will have an asterisk (*) next to their names in the script list.
- **Soft Reload Tool Script**: If the selected script is a
  :ref:`tool <doc_running_code_in_the_editor>`, reloads the script to execute it again.

> - **Copy Script Path**: Copies the currently selected script's relative path inside

Reviewer: ```suggestion
- **Copy Script Path**: Copies the currently selected script's relative path in
```

  the project using the ``res://`` prefix.
- **Show in FileSystem**: Finds and highlights the selected file in the FileSystem
  dock.
- **History Previous**: Changes the active script to the one that was previously
  opened. This is useful when you have multiple scripts open and want to quickly navigate
  back to the last script you were editing.
- **History Next**: After using `History Previous` to go back to an earlier script,
  this feature allows you to move forward through the script history, switching to

>   scripts that were previously accessed.

Author: Done

- **Theme**: Provides options to import an existing theme, save or reload it. Changing 
  theme settings is performed via `Editor Settings`.
- **Close**: Closes the active script.
- **Close All**: Closes all opened scripts and asks to save if there are unsaved changes.

> - **Close Other Tabs**: Closes all opened scripts except the selected one.

Reviewer: ```suggestion
- **Close Other Tabs**: Closes all open scripts except the selected one.
```

- **Close Docs**: Closes the class reference documentation pages, leaving only the

>   scripts opened.

Reviewer: ```suggestion
  scripts.
```

  current line.
- **Evaluate Selection***: Computes the values of the selected text if it contains `only`
  a mathematical expression, such as ``83 * 3`` or ``pow(2,3)``.
- **Toggle Word Wrap**: Disables the horizontal scrollbar by wrapping the long lines to
  the next line. Note that this is just a visual change and no new linebreaks are added.
- **Line**: Provides a set of line operations. Depending on the opened file, the options
  might also be directly in the Edit menu, instead of a submenu.

  - `Move Up`: Moves the current line or the selected line(s) one line up.
  - `Move Down`: Moves the current line or the selected line(s) one line down.
  - `Indent*`: Indents the text from the caret or the selected line(s), following the
    indentation setting.
  - `Unindent*`: Unindents the text from the caret or the selected line(s), following the
    indentation setting.
  - `Delete Line`: Deletes the current line or the selected line(s).
  - `Toggle Comment*`: Comments and uncomments the current line or the selected line(s).
    You can perform the same action by selecting line(s) and choosing the same action

>     after right-clicking on the selected text.

Author: The items here are not the context menu items, but the main menu. Since it is the first time I am mentioning this "Toggle Comment", I thought that it make sense to keep it.

>     after right-clicking on the selected text.

Author: @tetrapod00 What do you think about this? Can we resolve this?

>     after right-clicking on the selected text.

Reviewer 2: I ended up merging this as-is.

- **Go to Function...**: Opens the function list to jump to. You can achieve the same result
  by typing in the filter methods bar in the script panel.
- **Go to Line...**: Jumps to the written line number in the code editor.
- **Bookmarks**: Contains actions for the bookmark functionality, which you can use to find
  your way through your code easier, such as an incomplete section. Bookmarked lines will
  have a blue bookmark symbol left of the line number.

  - `Toggle Bookmark*`: Adds and removes the bookmark on the line where the caret is. You can
    also right click on a line to achieve this.
  - `Remove All Bookmarks`: Removes all bookmarks in the open document.
  - `Go to Next Bookmark`: Jumps to the next bookmark in the open document.
  - `Go to Previous Bookmark`: Jumps to the previous bookmark in the open document.
  - `Bookmarks` menu will also contain the list of bookmarked lines, including their line
    number and displaying the partial content in that line.

- **Breakpoints**: Breakpoints are helpful while debugging your code. Similar to "Bookmarks"
  menu, this menu lets you add or remove breakpoints, navigate between them and directly

>   jump to a specific breakpoint. An easy way to add a breakpoint is hovering the blank area
>   left of a line number. The cursor will show a faded red circle. Clicking it will add a

Reviewer 3: ```suggestion
  jump to a specific breakpoint. An easy way to add a breakpoint is hovering over the blank area
  left of a line number. It will show a faded red circle. Clicking it will add a
```

It is shown in the blank area mentioned, not on the cursor.

  have a blue bookmark symbol left of the line number.

  - `Toggle Bookmark*`: Adds or removes the bookmark on the line where the caret is. You can
    also right click on a line to achieve this.
  - `Remove All Bookmarks`: Removes all bookmarks in the open document.
  - `Go to Next Bookmark`: Jumps to the next bookmark in the open document.
  - `Go to Previous Bookmark`: Jumps to the previous bookmark in the open document.
  - `Bookmarks` menu will also contain the list of bookmarked lines, including their line
    number and displaying the partial content in that line.

- **Breakpoints**: Breakpoints are helpful while debugging your code. Similar to "Bookmarks"
  menu, this menu lets you add or remove breakpoints, navigate between them and directly
  jump to a specific breakpoint. An easy way to add a breakpoint is hovering over the blank area
  left of a line number. It will show a faded red circle. Clicking it will add a
  breakpoint and the circle will stay there. Clicking on a circle removes the breakpoint.

**Debug** menu offers actions which can be used while debugging. See
:ref:`doc_debugger_tools_and_options` for more.

.. _doc_script_editor_coding_area:

Coding area

> -----------

Reviewer 2: Is this name used anywhere else, prior to this page existing?
And is there some other name used for it in the editor itself?

We have to be a little careful with giving things names.

Author: Unfortunately no. What do you suggest using for this section?

Reviewer 2: Let's:
- Double check if it has a specific name anywhere else in the docs, or in the engine itself
- Maybe ask in the dev chat if there is another existing name for it.
- *Consider* avoiding specifically referring to this part of the UI. There is a chance that the proper name for this might just be "script editor" or "text editor". since it is the *main* part of the Script Editor workspace. There might not actually be a name for it like there is for the viewport for 2D and 3D.
- Consider calling it the "coding area", but avoid calling the section header "Coding area". If we get the name wrong for a while in the official docs it's *fine*, but titles of sections are slightly harder to change.

Author: Thanks. I will leave it as it is now.

Reviewer 4: I would not hold up this PR for this issue, but I've never seen this called anything but "script editor".

Reviewer 2: I've decided this isn't worth changing (at this moment). Let's go ahead and merge.

this value is set to 100 characters, by default. Both values can be changed, or the
display of the line can be toggled in the "Appearance" settings of the text editor.

.. |override| image:: img/script_editor_icons/override.webp
.. |receiver| image:: img/script_editor_icons/receiver.webp
.. |foldable| image:: img/script_editor_icons/Foldable.webp

In the script, to the left of function definitions, you might see additional icons. The |override|
icon indicates that this function is an :ref:`override <doc_overridable_functions>` of an existing 
function. Clicking it opens the documentation of the original function. The |receiver| icon means 
that it is a receiving method of a signal. Clicking it shows where the signal is coming 
from. A |foldable| icon to the left of the line denotes a foldable block. You can 
click to collapse or expand it. 
Alternatively, the ellipsis (...) icon can also be clicked to expand a folded block.

The example below summarizes the paragraph above. Lines 52, 56, and 58 are foldable blocks,
line 57 is a code region with the name "New Code Region," which you can also fold, and line
62 is a folded block.

> Line 53 is a bookmark, which can quickly be jumped to using the "Go To > Bookmarks" menu.

Reviewer 2: ```suggestion
Line 53 is a bookmark, which can quickly be jumped to using the **Go To > Bookmarks** menu.
```

Line 55 is a breakpoint that can be used in :ref:`debugging <doc_overview_of_debugging_tools>`.

.. image:: img/script_editor_icons/text_editor_coding_area_indicators.webp

Many of the colors of the text editor such as highlighting colors, or even breakpoint or
bookmark icon colors can be customized. You can experiment them by opening the text editor

> settings navigating to `Editor > Editor Settings > Text Editor` group.

Reviewer 2: ```suggestion
settings navigating to **Editor > Editor Settings > Text Editor** section.
```
